### ISSUE TITLE
What is your average running time for React Testing Library tests?

### ISSUE THREAD

**Issue by kensongoo** [testing-library/react-testing-library#819]
We have just started using React Testing Library. The biggest issue that we are facing right now is the test takes very long to run. One simple integration test (testing the behavior of several components on a container) takes about 80s to 100s. The test running speed is equal or worse than our Cypress tests now. 

What are the recommended ways to troubleshoot the running speed of the react testing library?

Also, what are the main culprits of slowness? Could it be component re-rendering issues? 

What are your average test running speed?

**Comment by nickmccurdy** [comment]
JSDOM can be fairly slow, and some parts of testing library can be slow, but it's hard to tell without seeing the test. Could you post code for testing library and Cypress in comparison?

**Comment by kentcdodds** [comment]
I'll add to what @nickmccurdy said by saying that you might consider using https://github.com/bahmutov/cypress-react-unit-test + https://github.com/testing-library/cypress-testing-library and you'll get basically the same experience just running in a real browser. Hard to tell whether it'll be faster though. The overhead of loading a real browser and the challenge of running in parallel is pretty significant.

**Comment by domarmstrong** [comment]
Stick console.times throughout your test suite and you can get an idea of which are the slow parts.

I've found `ByRole` queries can be in the region of 40x slower than some simpler equivalent ByText etc queries (but not always), so we are trying to only use ByRole when the semantics are significantly better (which they are in a lot of cases annoyingly, but still not always necessary).

Make sure you are using fake timers and not waiting for any realtime timers.

Also if using jest and running suites in parallel, I've found setting the max workers to 50% knocked 20% of our test suite run time, that could depend on the hardware you are using though and wouldn't affect running a single suite.

**Comment by dougbacelar-houzz** [comment]
Perhaps you could try disabling the [isInaccessible checks](https://github.com/testing-library/dom-testing-library/blob/c6e7a83948ef0d1822c5df1980b133ceb7e70dca/src/queries/role.js#L125-L129) on local environments and only running it on CI.

If there is a specific component that takes too long to render and it is unrelated to your test, you could try mocking that. As long as it is tested somewhere else the tradeoff might be worth it.

**Comment by dougbacelar** [comment]
Perhaps you could try disabling the [isInaccessible checks](https://github.com/testing-library/dom-testing-library/blob/c6e7a83948ef0d1822c5df1980b133ceb7e70dca/src/queries/role.js#L125-L129) on local environments and only running it on CI. Unsure how much that would help though.

If there is a specific component that takes too long to render and it is unrelated to your test, you could try mocking that. As long as it is tested somewhere else the tradeoff might be worth it.

**Comment by dougbacelar** [comment]
You might want to check the comments in this thread:

https://github.com/testing-library/dom-testing-library/issues/820#issuecomment-727006528

**Comment by eps1lon** [comment]
Closing in favor of #820 where we discuss performance of `*ByRole` and potential workarounds. There is no useful answer to the average running time of your tests since that depends on the UI you render and what you test.

**Comment by lpaulger** [comment]
> #820

I'm running into poor performance with react-testing-library and `mock-service-worker` libraries and trying to dig into the reason.

@eps1lon When I got into #820 I see not mention of `*ByRole` being discussed, perhaps it's a different ticket?

**Comment by kgregory** [comment]
@lpaulger It's this [issue 820 (for dom-testing-library)](https://github.com/testing-library/dom-testing-library/issues/820#issuecomment-727006528).  @dougbacelar had linked it earlier, but when the issue was referenced in a subsequent comment by number it was linked to an issue in this repo.
